Fix overflow of shmCommittedGxactArray on standby

darthunix opened the pull request:
This PR fixes issue https://github.com/greenplum-db/gpdb/issues/9207.

GPDB stores global transactions in a XLOG checkpoint record, that
are unpacked by a standby and added to shmCommittedGxactArray while
replaying on recovery. It is possible to get a inconsistency between
QD and checkpointer:

1. QD: commit gid 1 <- or after checkpoint.redo but before
   getDtxCheckPointInfo()
2. CP: checkpoint.redo
3. CP: getDtxCheckPointInfo() get gid1 for checkpoint xlog
4. QD: forget gid 1
5. CP: checkpoint  <- when redoing it on standby, gid 1 is added
   to the distribution transaction list again.

As a result some global transactions can be added twice, which
finally caused the overflow of shmCommittedGxactArray.

In fact, global transactions contained in checkpoint XLOG record
can be replayed on a standby only once on startup and the following
can be skipped.

darthunix commented:
To run isolation test you need:
```
cp src/test/regress/regress/*.pl src/test/isolation2
cp src/test/regress/regress/*.pm src/test/isolation2
cd src/test/isolation2
make

./pg_isolation2_regress --init-file=init_file_isolation2 standby_gxact_overflow
```

dreamedcheng commented on src/backend/cdb/cdbdtxrecovery.c:
```diff
@@ -558,17 +561,38 @@ redoDistributedCommitRecord(TMGXACT_LOG *gxact_log)
+#ifdef FAULT_INJECTOR
+		if (SIMPLE_FAULT_INJECTOR("gxacts_overflow") == FaultInjectorTypeSkip)
+		{
+			gxacts_limit = 1;
```
I wonder if we can change the value of `max_tm_gxacts` directly here instead of introducing a new parameter?

darthunix replied:
I also thought of this idea but finally rejected it - it seems unsafe to modify a global variable `max_tm_gxacts` for fault injection purpose (anyway we have to restore it after injecting a fault).

darthunix commented:
@dreamedcheng [suggested](https://github.com/greenplum-db/gpdb/pull/11071#issuecomment-719097002) in his upstream PR to use `pg_stat_replication` on master to detect the case when standby fails instead of a python wrapper over `gstate`. It seems to be a good idea (and may be we should wait a little with this PR to see the final solution in master code).

Stolb27 commented on src/test/isolation2/sql/standby_gxact_overflow.sql:
```diff
@@ -0,0 +1,85 @@
+-- after try #2 we got an overflow on a standby.
+--
+-- start_ignore
+1: create table t (a int);
```
Relation with the same name already exists (was created and wasn't drop) in https://github.com/arenadata/gpdb/blob/35827cabc11f70da2db583b26da8f647ae04e0d4/src/test/isolation2/sql/segwalrep/failover_with_many_records.sql#L12

darthunix replied:
Fixed

darthunix commented:
I have force pushed https://github.com/arenadata/gpdb/commit/35827cabc11f70da2db583b26da8f647ae04e0d4 with https://github.com/arenadata/gpdb/pull/125/commits/5cf6f49ca3b1d2a4fa727c70dc5a4b6cefd41818. Changes (thanks to ideas from @dreamedcheng and @Stolb27 :
1. Removed [standby_is_down()](https://github.com/arenadata/gpdb/commit/35827cabc11f70da2db583b26da8f647ae04e0d4#diff-554dea360b269b2fe38ea44630f1510df0d5651a562d5196c3e677a948e58cb9R14-R26) function. It was implemented as a `plpythonu` wrapper over `gpstate`. But a much more simple solution to rely on [pg_stat_replication](https://github.com/arenadata/gpdb/pull/125/commits/5cf6f49ca3b1d2a4fa727c70dc5a4b6cefd41818#diff-554dea360b269b2fe38ea44630f1510df0d5651a562d5196c3e677a948e58cb9R71) - there is a single replication entry when standby is up and empty if standby is down.
1. [Speeded up](https://github.com/arenadata/gpdb/pull/125/commits/5cf6f49ca3b1d2a4fa727c70dc5a4b6cefd41818#diff-554dea360b269b2fe38ea44630f1510df0d5651a562d5196c3e677a948e58cb9R18) a test for gpdb built with assert.
1. [Recover](https://github.com/arenadata/gpdb/pull/125/commits/5cf6f49ca3b1d2a4fa727c70dc5a4b6cefd41818#diff-554dea360b269b2fe38ea44630f1510df0d5651a562d5196c3e677a948e58cb9R74) standby (if it failed).
1. [Renamed](https://github.com/arenadata/gpdb/pull/125/commits/5cf6f49ca3b1d2a4fa727c70dc5a4b6cefd41818#diff-554dea360b269b2fe38ea44630f1510df0d5651a562d5196c3e677a948e58cb9R14) test table

darthunix commented:
I have backported https://github.com/greenplum-db/gpdb/pull/11071 to our 6X branch as @dreamedcheng 's commit was strongly inspired by this branch. Though https://github.com/greenplum-db/gpdb/pull/11071 has a failed `expand_table` test it is an inaccuracy in master branch (the output error message ` is not a table or foreign table` has changed since 6X to `is not a table, materialized view, or foreign table` but was not fixed in the test). 6X is not affected (I have run this test locally).

dreamedcheng committed:
Fix overflow of shmCommittedGxactArray on standby (#11071)

Previously, standby will replay checkpoint XLOG record's DTX info
in function XLogProcessCheckpointRecord. However, is some certain cases,
it will cause anomaly: When a DTX has flushed FORGET COMMITTED XLOG into
disk, but didn't change its own state to DTX_STATE_INSERTED_FORGET_COMMITTED.
If at this very moment, checkpoint process is calculating DTX info, it
will include the DTX into its XLOG record. So when standby replaying this
checkpoint XLOG record from master, it will add an already forgotten GID to
shmCommittedGxactArray again, which may cause the overflow of shmCommittedGxactArray.

Since DTX info stored in checkpoint XLOG record has been populated earlier in
ReadCheckpointRecord(), there is no need to call XLogProcessCheckpointRecord()
again during recovery.

A backport of afcf30bedb3f1a2741cd79b98f232b10be7f65b4

Co-authored-by: wuchengwen <wcw190496@alibaba-inc.com>
Co-authored-by: Denis Smirnov <sd@arenadata.io>

darthunix commented:
Finally https://github.com/greenplum-db/gpdb/pull/11071 have been merged to `master` branch. So this force push is a backport of https://github.com/greenplum-db/gpdb/commit/afcf30bedb3f1a2741cd79b98f232b10be7f65b4 to `adb-6.x`.

darthunix commented:
Pivotal also [backported](https://github.com/greenplum-db/gpdb/commit/b81a580a39db8f09ec1966d11258451230c7f58a) this fix to 6X - I suggest to close this PR. What is your opinion, @Stolb27 ?
